torch/csrc/api/include/torch/data/ctf/ctf_parser.h

```diff
@@ -0,0 +1,931 @@
+      CTFSequenceData sequence;
+
+      for (auto const& stream : dataset_->input_streams_info) {
+        CTFInputStreamID input_stream_id =
+            dataset_->input_streams_map[stream.name];
+        if (stream.storage == CTFDataStorage::Dense) {
+          // TODO: Performance consideration: CNTK knows the number of samples
+          // in the sequence, allowing accurate memory reservation (index built
+          // during init)
+          sequence.emplace_back(
+              std::make_shared<CTFDenseInputStreamData<DataType>>(
+                  input_stream_id, stream.dimension));
+        } else {
+          sequence.emplace_back(
+              std::make_shared<CTFSparseInputStreamData<DataType>>(
+                  input_stream_id, stream.dimension));
+        }
+      }
+
+      dataset_->sequences.emplace_back(sequence);
+    }
+
+    // Reads the Input Stream name and lookup its input stream reference
+    CTFInputStreamID input_stream_id;
+    if (!get_input_stream_name(input_stream_id)) {
+      return false;
+    }
+    const CTFInputStreamInformation& input_stream =
+        dataset_->input_streams_info[input_stream_id];
+
+    // Appends all values to the input stream
+    if (!get_input_stream(input_stream)) {
+      return false;
+    }
+
+    // TODO: Check actual number of values records of the stream
+    return true;
+  }
+
+  // Parses input name from buffer and returns both CTFInputStreamInformation
+  // reference and true if the input name belongs to an existing Input Stream
+  bool get_input_stream_name(CTFInputStreamID& input_stream_id) {
+#ifdef CTF_DEBUG
+    // For logging purposes
+    size_t initial_pos = reader_->get_position();
+#endif
+    // idx will be used to iterate through scratch_ for local string parsing
+    size_t idx = 0;
+
+    // CTF Name must start with a '|'
+    if (!is_name_prefix(reader_->peek_char())) {
+#ifdef CTF_DEBUG
+      std::cout << "Not a CTF Name at position " << initial_pos << std::endl;
+#endif
+      return false;
+    }
+
+    // Discard | and get all consecutive digits and alpha characters
+    char c = reader_->get_char();
+    while (is_digit(reader_->peek_char()) || is_alpha(reader_->peek_char())) {
+      c = reader_->get_char();
+      scratch_[idx++] = c;
+    }
+    scratch_[idx] = '\0';
+
+    // Discard delimiters after the CTF Name
+    while (is_value_delimiter(reader_->peek_char())) {
+      c = reader_->get_char();
+    }
+
+    // After CTF Name, there must be a CTF value or another CTF Name
+    c = reader_->peek_char();
+    if (!is_number(c) && !is_name_prefix(c) && !is_eol(c)) {
+#ifdef CTF_DEBUG
+      std::cerr << "Unexpected symbol '" << c << "' after CTF Name at position "
+                << reader_->get_position() << std::endl;
+#endif
+      reader_->rewind_char();
+      return false;
+    }
+
+    // Return the CTF Name
+    // TODO: Can be done better?
+    std::string name = std::string(scratch_.begin(), scratch_.begin() + idx);
```

Comment:
I think we don't necessary need to create a new string here. Just using char* name = scratch_.data() would suffice.